drivers/mm/mm_drv_intel_adsp_mtl_tlb.c

```diff
@@ -672,29 +682,182 @@ static int sys_mm_drv_mm_init(const struct device *dev)
 	 * find virtual address range which are unused
 	 * in the system
 	 */
-	uintptr_t unused_l2_start_aligned =
-		ROUND_UP(POINTER_TO_UINT(&unused_l2_sram_start_marker),
-					 CONFIG_MM_DRV_PAGE_SIZE);
-
-	if (unused_l2_start_aligned < L2_SRAM_BASE ||
-	    unused_l2_start_aligned > L2_SRAM_BASE + L2_SRAM_SIZE) {
+	if (L2_SRAM_BASE + L2_SRAM_SIZE < UNUSED_L2_START_ALIGNED ||
+	    L2_SRAM_BASE > UNUSED_L2_START_ALIGNED) {
 		__ASSERT(false,
 			 "unused l2 pointer is outside of l2 sram range %p\n",
-			 unused_l2_start_aligned);
+			 UNUSED_L2_START_ALIGNED);
 		return -EFAULT;
 	}
 
 	/*
 	 * Unmap all unused physical pages from the entire
 	 * virtual address space to save power
 	 */
-	size_t unused_size = CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE -
-			     unused_l2_start_aligned;
+	size_t unused_size = L2_SRAM_BASE + L2_SRAM_SIZE - UNUSED_L2_START_ALIGNED;
 
-	ret = sys_mm_drv_unmap_region(UINT_TO_POINTER(unused_l2_start_aligned),
+	ret = sys_mm_drv_unmap_region(UINT_TO_POINTER(UNUSED_L2_START_ALIGNED),
 				      unused_size);
 
 	return 0;
 }
 
-SYS_INIT(sys_mm_drv_mm_init, POST_KERNEL, 0);
+#ifdef CONFIG_PM_DEVICE
+static void adsp_mm_save_context(void *storage_buffer)
+{
+	/* at this point the system must be almost stopped
+	 * as this don't change anything RAM, there's no cache invalidation with write back,
+	 * assuming the dcache memory had been invalidated before
+	 * this procedure will store all mapped memory to given storage pointer
+	 * at this point the memory is still up&running so its safe to use libraries like memcpy
+	 */
+	uint16_t entry;
+	uint32_t entry_idx;
+	int page_idx;
+	uint32_t phys_addr;
+	volatile uint16_t *tlb_entries = UINT_TO_POINTER(TLB_BASE);
+	uint8_t *location = (uint8_t *) storage_buffer;
+
+	/* first, store the existing TLB */
+	memcpy(location, UINT_TO_POINTER(TLB_BASE), TLB_SIZE);
+	location += TLB_SIZE;
+
+	/* save context of all the pages that had been mapped */
+	/* FIXME - currently saving the memory above the unused marker only
+	 * according to needs there may be more memory saved, even the whole HPSRAM (start_page=0)
+	 */
+	uint32_t start_page = (UNUSED_L2_START_ALIGNED -  POINTER_TO_UINT(L2_SRAM_BASE))
+					/ CONFIG_MM_DRV_PAGE_SIZE;
+
+	for (page_idx = start_page; page_idx < L2_SRAM_PAGES_NUM; page_idx++) {
```

Comment:
This loop will go only in physical memory range while TLB table is for virtual range. Now it will loose entries for virtual addresses higher than L2_SRAM_BASE + (L2_SRAM_PAGES_NUM * CONFIG_MM_DRV_PAGE_SIZE). 
It should go from CONFIG_KERNEL_VM_BASE to (CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE)

Reply:
that's not true.
see line 743 - there's a remap of used physical memory pages 1:1 to virtual space, so I don't care about how the physical pages used to be mapped to a virtual space. Its enough to store all physical memory - mapped 1:1 to virtual!

This is much faster than checking all virtual mapping and taking care for regions mapped to several addresses etc.